distributed/tests/test_client.py

```diff
@@ -7503,3 +7504,73 @@ async def test_wait_for_workers_updates_info(c, s):
     async with Worker(s.address):
         await c.wait_for_workers(1)
         assert c.scheduler_info()["workers"]
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_client_story(c, s, *workers):
+    f = c.submit(inc, 1, workers=workers[0].name)
+    assert await f == 2
+    story = await c.story(f.key)
+
+    assert_story(
+        story,
+        [
+            (f.key, "released", "waiting", {f.key: "processing"}),
+            (f.key, "waiting", "processing", {}),
+            (f.key, "processing", "memory", {}),
+            (f.key, "compute-task"),
+            (f.key, "released", "waiting", "waiting", {f.key: "ready"}),
+            (f.key, "waiting", "ready", "ready", {f.key: "executing"}),
+            (f.key, "ready", "executing", "executing", {}),
+            (f.key, "put-in-memory"),
+            (f.key, "executing", "memory", "memory", {}),
+        ],
+        ordered_timestamps=False,
+    )
```

Comment:
@mrocklin Here is the test case including both scheduler and worker events. Testing of stimuli remains to be added.

Reply:
In general I dislike the assert_story tests.  They seem really brittle to me.  This also doesn say to me "this closes the circle".  Does it say that to you?  If we failed in closing the circle is there something that we would immediately notice without having to understand this code exactly?

Reply:
Or maybe I should ask "what does it mean to close the circle?" Can we define that somehow?  Currently this test says to me "this task transitions from released to waiting and then to processing and then to memory and then to compute-task and then to waiting and then to ready and then to executing and then to memory" which doesn't actually tell me much.  I view tests like these as more of a liability (I'm afraid that I'll have to spend a lot of time in the future to understand it when touching some other part of the code) rather than an asset (something that would tell me "this closes the circle").

Reply:
Ah, the confusion might be that I haven't pushed up the stimulus_id comparisons. Just testing locally. 

Would that close the circle do you think?

Reply:
For example, to me it means that we can trace any event downstream to a root stimulus.  By "trace" I mean that I can connect stimulus IDs to other stimulus IDs until I get back to some root source. 

I don't care about the values of those stimulus IDs (and so I wouldn't want them in the test) but I would want to create some function like 

```python
def can_trace_to_source(...):
    ...
```

And apply it to various events and verify that they cross from the scheduler and worker coherently. I absolutely wouldn't want to test the exact story that they tell (because this is certainly going to change over the next few weeks, and I want my test to provide value for more than a few weeks).

I don't know stimulus IDs well enough to say that this is correct though.  I'm hoping that you do.  I think that we need to lift up a little bit from highly specific "this is exactly what happened" tests to something that tests a broader invariant.

I'm also happy to switch to a live call tomorrow if that's easier.  I suspect that we're speaking past each other a little here.

Reply:
>  And apply it to various events and verify that they cross from the scheduler and worker coherently. 

> I'm also happy to switch to a live call tomorrow if that's easier. I suspect that we're speaking past each other a little here.

Thanks, I'd appreciate that, I've been mostly testing with the existing story utilities in mind. It'll also help me to think about your suggestion with a fresh head.